File: Graph.java
import java.util.*;

public class Graph<K, V>
{
	
	private Boolean directed;
	private HashMap<K, V> vertices;
	private HashMap<K, LinkedList<Edge<K>>> adj_list;
	
	public Graph(Boolean directed)
	{
		__init__();
		this.directed = directed;
	}
	
	public Graph()
	{
		__init__();
		this.directed = false;
	}
	
	private void __init__()
	{
		this.vertices = new HashMap<K, V>();
		this.adj_list  = new HashMap<K, LinkedList<Edge<K>>>();
	}
	
	protected LinkedList<Edge<K>> get_neighbors(K key)
	{
		if(!vertices.containsKey(key)) throw new IllegalArgumentException(
				"Error: Vertex " + key + " Does Not Exist.");
		
		return this.adj_list.get(key);
	}
	
	public void add_edge(K key1, V vertex1, K key2, V vertex2, double weight)
	{
		if(!this.vertices.containsKey(key1)) add_vertex(key1, vertex1);
		if(!this.vertices.containsKey(key2)) add_vertex(key2, vertex2);
		
		Edge<K> key_1_2 = new Edge<K>(key2, weight);
		this.adj_list.get(key1).add(key_1_2);
		
		if(this.directed == false)
		{
			Edge<K> key_2_1 = new Edge<K>(key1, weight);
			this.adj_list.get(key2).add(key_2_1);			
		}
	}
	
	private void add_vertex(K key, V vertex)
	{
		this.vertices.put(key, vertex);
		LinkedList<Edge<K>> list = new LinkedList<Edge<K>>();
		this.adj_list.put(key, list);
	}
	
	//Depth First Search Variables:
	private HashMap<K, Boolean> dfs_visited;
	private HashMap<K, Integer> dfs_finishTimes;
	private int dfsTime;
	
	public LinkedList<K> topological_sort()
	{
		dfs_visited = new HashMap<K, Boolean>();
		dfs_finishTimes = new HashMap<K, Integer>();
		LinkedList<K> topList = new LinkedList<K>();
		
		for(K key : this.vertices.keySet())
		{
			this.dfs_visited.put(key, false);
			this.dfs_finishTimes.put(key,  -1);
		}
		
		for(K key : this.vertices.keySet())
		{
			if(this.dfs_visited.get(key) == false)
			{
				dfs_visit(topList, key);
			}
		}
		
		return topList;
	}
	
	private void dfs_visit(LinkedList<K> topList, K iKey)
	{
		this.dfs_visited.put(iKey, true);
		
		LinkedList<Edge<K>> neighborEdges = this.adj_list.get(iKey);
		for(Edge<K> edge : neighborEdges)
		{
			K edgeKey = edge.getKey();
			if(this.dfs_visited.get(edgeKey) == false)
			{
				dfs_visit(topList, edgeKey);
			}
		}
		
		this.dfsTime++;
		this.dfs_finishTimes.put(iKey, this.dfsTime);
		//Add this vertex to the FRONT of the list:
		topList.addFirst(iKey);	
		
	}
	
	public void print_graph()
	{
		for(K key: vertices.keySet())
		{
			System.out.println(key + ":  ");
			List<Edge<K>> edgeList = this.adj_list.get(key);
			for(Edge<K> e : edgeList)
			{
				System.out.println("\t" + e.getKey() + ", " + e.getWeight());
			}
			System.out.println("");
		}
	}
	
	//__________________________________________________
	//* * * * SHORTEST PATH PROPERTIES * * * * *
	HashMap<K, Double> distanceTable;
	HashMap<K, K> predecessorTable;
	
	private class EdgeComparator implements Comparator <Edge<K>> 
	{
		public int compare(Edge<K> o1, Edge<K> o2) 
		{
			return (o1.getWeight() < o2.getWeight()) ? 0 : 1;
		}	
	}
	
	public Stack<K> shortestPath(K from, K to)
	{
		Comparator<Edge<K>> comp = new EdgeComparator();
		
		//Build distance & predecessor tables:
		dijkstra_sp(from, comp);
		
		Stack<K> path = new Stack<K>();
		K previous = to;
		while(previous != from && previous != null)
		{
			path.push(previous);
			previous = predecessorTable.get(previous);
		}
		if(previous == null) return null;
		path.push(previous);
		return path;
	}
	
	private void dijkstra_sp(K from, Comparator<Edge<K>> comp)
	{
		
		distanceTable = new HashMap<K, Double>();
		predecessorTable = new HashMap<K, K>();
		HashMap<K, Boolean> visited = new HashMap<K, Boolean>();
		
		//Initialize Queue that keeps "light edge" at head 
		PriorityQueue<Edge<K>> queue = new PriorityQueue<Edge<K>>(comp);
		Edge<K> source_edge = new Edge<K>(from, 0);
		queue.add(source_edge);
		
		for(K key : this.vertices.keySet())
		{
			distanceTable.put(key, Double.MAX_VALUE);
			predecessorTable.put(key, null);
			visited.put(key, false);
		}
	
		visited.put(from, true);
		distanceTable.put(from, (double)0);
		
		while(queue.isEmpty() == false)
		{
			K onKey = queue.poll().getKey();
			double onDistance = distanceTable.get(onKey);
					
			LinkedList<Edge<K>> neighborEdges = adj_list.get(onKey);
			
			for(Edge<K> edge : neighborEdges)
			{
				//Check if neighbor just found a shorter path:
				K neighborKey = edge.getKey();
				if(distanceTable.get(neighborKey) > (onDistance + edge.getWeight()))
				{
					predecessorTable.put(neighborKey, onKey);
					distanceTable.put(neighborKey, onDistance + edge.getWeight());		
				}
				
				if(visited.containsKey(neighborKey) == false) queue.add(edge);
				
			}	
		}
	}//End Dijkstra (Build distance & predecessor tables)
	
	
}

File: Graph.py
import heapq
import sys

class Vertex:
    def __init__(self, data, index):
        self.data = data
        self.index = index

class Graph:

    def __init__(self, undirected=True):
        self.size = 0
        self.vList = []
        self.vMap = {}
        self.adjList = []
        self.undirected = undirected
        self.weights = {}

    def addEdge(self, v1, v2, weight):
        if v1 not in self.vMap:
            i1 = self.size
            self.size += 1
            self.addVertex(v1, i1)
        else: i1 = self.vMap[v1].index
        if v2 not in self.vMap:
            i2 = self.size
            self.size += 1
            self.addVertex(v2, i2)
        else: i2 = self.vMap[v2].index

        self.adjList[i1].add(i2)
        edge = str(i1) + "," + str(i2)
        self.weights[edge] = weight

        if(self.undirected):
            self.adjList[i2].add(i1)
            edge = str(i2) + "," + str(i1)
            self.weights[edge] = weight

    def addVertex(self, vData, index):
        vertex = Vertex(vData, index)
        self.adjList.append(set([]))
        self.vList.append(vertex)
        self.vMap[vData] = vertex

    def getNeighbors(self,v):
        if v<0 or v>=self.size:
            print('error: v index out of bounds')
            return set()
        return self.adjList[v]

    def getEdgeWeight(self,v1,v2):
        if v1 < 0 or v2 < 0:
            print('error: v index out of bounds')
            return 0
        edge = str(v1) + "," + str(v2)
        return self.weights[edge]

    def printGraph(self):
        print('Number of Nodes = ', self.size)
        for vertex in self.vList:
            data = vertex.data
            vId = vertex.index
            neighbors = self.getNeighbors(vId)
            print(data, ' Connections = ')
            for n in neighbors:
                neighbor = self.vList[n]
                print('   ', neighbor.data)
            print()

    #Depth first search:
    def dfs(self):
        time = 0
        visited = set()
        topList = []
        for v in self.vList:
            if v.index not in visited:
                self.dfsVisit(v.index, visited, topList)
        return topList
    def dfsVisit(self, vertexId, visited, topList):
        visited.add(vertexId)
        neighbors = self.getNeighbors(vertexId)
        for n in neighbors:
            if n not in visited:
                self.dfsVisit(n, visited, topList)
        topList.append(vertexId)

    #Breadth first search
    def bfs(self, startVertexId):
        bfsList = []
        bfsVisited = set([])
        bfsList = self.bfsVisit(startVertexId, bfsVisited, bfsList)
        return bfsList
    def bfsVisit(self, vertexId, bfsVisited, bfsList):
        bfsVisited.add(vertexId)
        bfsList.append(vertexId)
        neighbors = self.getNeighbors(vertexId)
        for n in neighbors:
            if n not in bfsVisited:
                self.bfsVisit(n, bfsVisited, bfsList)
        return bfsList

    #Dijkstra's single source shortest path algorithm
    def buildShortestPathTable(self, source):
        distanceTable = {}
        predecessorTable = {}
        for i in range(len(self.vList)):
            distanceTable[i] = sys.maxsize
            predecessorTable[i] = 'na'

        distanceTable[source] = 0
        predecessorTable[source] = 'na'
        visited = set()
        beenQueued = set()
        queue = [(0, source)]
        beenQueued.add(source)
        while len(queue) > 0:
            distInfo = heapq.heappop(queue)
            v = distInfo[1]
            vDist = distInfo[0]
            visited.add(v)
            neighbors = self.getNeighbors(v)
            for n in neighbors:
                if n not in visited:
                    nDist = vDist + self.getEdgeWeight(v,n)
                    currentDistance = distanceTable[n]
                    if nDist < currentDistance:
                        predecessorTable[n] = v
                        distanceTable[n] = nDist
                        #Update queue distance:
                        if n in beenQueued:
                            queue.remove((currentDistance, n))
                            heapq.heappush(queue, (nDist, n))

                    if n not in beenQueued:
                        heapq.heappush(queue, (nDist, n))

        self.distanceTable = distanceTable
        self.predecessorTable = predecessorTable

    #Returns the shortest path from source to dest vertices (if a path exists).
    #Dijkstra's algorithm DOES NOT work with negative edges.
    def shortestPath(self,source, dest, buildTable=True):
        if source not in self.vMap or dest not in self.vMap:
            return 'Path does not exist.'
        dest = self.vMap[dest].index #Get adjList index of destination
        source = self.vMap[source].index #Get adjList index of source
        if buildTable:
            self.buildShortestPathTable(source)
        stack = [self.vList[dest].data]
        currentVertex = self.predecessorTable[dest]
        while currentVertex != source:
            v = self.vList[currentVertex].data
            stack.insert(0, v)
            currentVertex = self.predecessorTable[currentVertex]
        stack.insert(0, self.vList[source].data)
        return stack


    #Prim's Algorithm
    # @return: A new Graph object that represents
    #           a minimum spanning tree for this graph
    def minSpanTree(self):
        if self.size == 0:
            return Graph(undirected=True)

        mst = Graph(undirected=True)

        s = 0
        edgeQueue = []
        inMST = set([s])
        beenQueued = set([])

        sNeighbors = self.getNeighbors(s)
        for n in sNeighbors:
            w = self.getEdgeWeight(s, n)
            edge = (w, (s,n))
            heapq.heappush(edgeQueue, edge)
            beenQueued.add(edge)

        while len(edgeQueue) > 0:
            edge = heapq.heappop(edgeQueue)
            weight = edge[0]
            v1 = edge[1][0]
            v2 = edge[1][1]

            #Check if this should be skipped.
            if v1 in inMST and v2 in inMST: continue

            v1d = self.vList[v1].data
            v2d = self.vList[v2].data
            mst.addEdge(v1d,v2d,weight)
            if v1 not in inMST: inMST.add(v1)
            if v2 not in inMST: inMST.add(v2)

            if len(inMST) == self.size:
                return mst

            neighbors = self.getNeighbors(v2)
            for n in neighbors:
                w = self.getEdgeWeight(v2, n)
                nEdge = (w, (v2,n))
                if nEdge not in beenQueued:
                    heapq.heappush(edgeQueue, nEdge)
                    beenQueued.add(nEdge)

    #Returns the sum of all edge-weights
    def getTreeWeight(self):
        totalWeight = 0
        counted = set()
        for i in range(len(self.adjList)):
            for n in self.adjList[i]:
                if (i,n) in counted:
                    continue
                w = self.getEdgeWeight(i, n)
                totalWeight += w
                counted.add((i,n))
                if self.undirected:
                    counted.add((n,i))
        return totalWeight
